File: webauthn-server-core/src/main/scala/com/yubico/webauthn/impl/ExtensionsValidation.scala
package com.yubico.webauthn.impl

import com.upokecenter.cbor.CBORObject
import com.yubico.scala.util.JavaConverters._
import com.yubico.webauthn.data.AuthenticationExtensionsClientInputs
import com.yubico.webauthn.data.PublicKeyCredential
import com.yubico.webauthn.data.AuthenticatorResponse
import com.yubico.webauthn.util.WebAuthnCodecs

import scala.collection.JavaConverters._


object ExtensionsValidation {

  def validate(requested: Option[AuthenticationExtensionsClientInputs], response: PublicKeyCredential[_ <: AuthenticatorResponse]): Boolean = {
    val requestedExtensionIds: Set[String] = requested map { _.fieldNames.asScala.toSet } getOrElse Set.empty

    val clientExtensionIds: Set[String] = response.getClientExtensionResults.fieldNames.asScala.toSet
    assert(
      clientExtensionIds subsetOf requestedExtensionIds,
      s"Client extensions {${clientExtensionIds.toSeq.sorted.mkString(", ")}} are not a subset of requested extensions {${requestedExtensionIds.toSeq.sorted.mkString(", ")}}."
    )

    val authenticatorExtensionIds: Set[String] =
      response.getResponse.getParsedAuthenticatorData.getExtensions.asScala
        .map(_.getKeys.asScala.toSet.map((o: CBORObject) => o.AsString()))
        .getOrElse(Set.empty)
    assert(
      authenticatorExtensionIds subsetOf requestedExtensionIds,
      s"Authenticator extensions {${authenticatorExtensionIds.toSeq.sorted.mkString(", ")}} are not a subset of requested extensions {${requestedExtensionIds.toSeq.sorted.mkString(", ")}}."
    )

    true
  }

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/internal/ExtensionsValidation.java
package com.yubico.webauthn.internal;

import com.fasterxml.jackson.databind.JsonNode;
import com.upokecenter.cbor.CBORObject;
import com.yubico.internal.util.StreamUtil;
import com.yubico.webauthn.data.AuthenticatorResponse;
import com.yubico.webauthn.data.PublicKeyCredential;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.experimental.UtilityClass;


@UtilityClass
public class ExtensionsValidation {

    public static boolean validate(Optional<JsonNode> requested, PublicKeyCredential<? extends AuthenticatorResponse> response) {
        if (requested.isPresent() && !requested.get().isObject()) {
            throw new IllegalArgumentException(String.format(
                "Requested extensions must be a JSON object, was: %s",
                requested.get()
            ));
        }

        Set<String> requestedExtensionIds = requested.map(req -> StreamUtil.toSet(req.fieldNames())).orElseGet(HashSet::new);
        Set<String> clientExtensionIds = StreamUtil.toSet(response.getClientExtensionResults().fieldNames());

        if (!requestedExtensionIds.containsAll(clientExtensionIds)) {
            throw new IllegalArgumentException(String.format(
                "Client extensions {%s} are not a subset of requested extensions {%s}.",
                String.join(", ", clientExtensionIds),
                String.join(", ", requestedExtensionIds)
            ));
        }

        Set<String> authenticatorExtensionIds = response.getResponse().getParsedAuthenticatorData().getExtensions()
            .map(extensions -> extensions.getKeys().stream()
                .map(CBORObject::AsString)
                .collect(Collectors.toSet())
            )
            .orElseGet(HashSet::new);

        if (!requestedExtensionIds.containsAll(authenticatorExtensionIds)) {
            throw new IllegalArgumentException(String.format(
                "Authenticator extensions {%s} are not a subset of requested extensions {%s}.",
                String.join(", ", authenticatorExtensionIds),
                String.join(", ", requestedExtensionIds)
            ));
        }

        return true;
    }

}
